ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -55,60 +56,60 @@ where
+        let mut u = u.into_iter();
+        let mut v = v.into_iter();
+
+        debug_assert_eq!(u.len() % λ::USIZE, 0); // We should pad with zeroes eventually
 
-        let s = self.u.len() / λ::USIZE;
+        let s = u.len() / λ::USIZE;
 
         assert!(
             s > 1,
             "When the output is this small, you should call `compute_final_proof`"
         );
 
-        let mut next_proof_generator = ProofGenerator {
-            u: Vec::<F>::with_capacity(s),
-            v: Vec::<F>::with_capacity(s),
-        };
-
         let denominator = CanonicalLagrangeDenominator::<F, λ>::new();
-        let lagrange_table_r = LagrangeTable::<F, λ, U1>::new(&denominator, &r);
         let lagrange_table = LagrangeTable::<F, λ, <λ as Sub<U1>>::Output>::from(denominator);
-        let extrapolated_points = (0..s).map(|i| {
-            let start = i * λ::USIZE;
-            let end = start + λ::USIZE;
-            let p = &self.u[start..end];
-            let q = &self.v[start..end];
-            let p_extrapolated = lagrange_table.eval(p);
-            let q_extrapolated = lagrange_table.eval(q);
-            let p_r = lagrange_table_r.eval(p)[0];
-            let q_r = lagrange_table_r.eval(q)[0];
-            next_proof_generator.u.push(p_r);
-            next_proof_generator.v.push(q_r);
-            // p.into_iter() has elements that are &F
-            // p_extrapolated.into_iter() has elements that are F
-            // So these iterators cannot be chained.
-            zip(p, q)
-                .map(|(a, b)| *a * *b)
-                .chain(zip(p_extrapolated, q_extrapolated).map(|(a, b)| a * b))
-        });
-        let proof = ZeroKnowledgeProof::new(extrapolated_points.fold(
-            GenericArray::<F, TwoNMinusOne<λ>>::generate(|_| F::ZERO),
-            |acc, pts| zip(acc, pts).map(|(a, b)| a + b).collect(),
-        ));
-        (proof, next_proof_generator)
+        let mut p = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        let mut q = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        let mut proof: GenericArray<F, TwoNMinusOne<λ>> = GenericArray::generate(|_| F::ZERO);
+        for _ in 0..s {
+            for i in 0..λ::USIZE {
+                let x = u.next().unwrap_or(F::ZERO);
+                let y = v.next().unwrap_or(F::ZERO);
+                p[i] = x;
+                q[i] = y;
+                proof[i] += x * y;
+            }
+            let p_extrapolated = lagrange_table.eval(&p);
+            let q_extrapolated = lagrange_table.eval(&q);
+
+            for (i, (x, y)) in
+                zip(p_extrapolated.into_iter(), q_extrapolated.into_iter()).enumerate()
+            {
+                proof[λ::USIZE + i] += x * y;
+            }
+        }
+        ZeroKnowledgeProof::new(proof)
     }
 
-    pub fn compute_final_proof<λ: ArrayLength>(
-        &self,
+    pub fn compute_final_proof<λ: ArrayLength, I, J>(
+        u: I,
+        v: J,
         p_0: F,
```

Comment:
are these the masks?

Reply:
Yes. The random weights used for p(0) and q(0) come from PRSS and I'm trying to NOT have any dependency on that module from this code. As such, I'm assuming the caller will invoke PRSS to obtain random field elements and then send them in.